File: go/src/medium/61.rotate-list.go
package medium

import . "common"

func Rotate(head *ListNode, k int) *ListNode {
    if head == nil {
        return nil
    }
    copyHead := head

    length := 1

    for copyHead.Next != nil {
        length++
        copyHead = copyHead.Next
    }
    copyHead.Next = head

    for i := length - k % length; i > 1; i-- {
        head = head.Next
    }
    copyHead = head.Next
    head.Next = nil
    return copyHead
}

File: archive/python/old/61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        fast, slow = head, head
        size = 1
        # for fast.next = head
        while fast.next:
            size += 1
            fast = fast.next
        for _ in range(1, (size - k % size), -1):
            slow = slow.next
        fast.next = head
        head = slow.next
        slow.next = None
        return head
# @lc code=end

